### ISSUE TITLE
[Discussion] My experience using CsCheck in floating point heavy tests

### ISSUE THREAD

**Issue by SamPruden** [AnthonyLloyd/CsCheck#13]
I'm trying out CsCheck for some code that's very floating point heavy. (It's a 2D collision query engine). Here's my experience so far.

I'm loving the ergonomics of the library, but my initial findings are that it's pretty weak on certain classes of numerical tests. I can trivially construct some common failure patterns which will almost never be caught, and when they are caught, will not be shrunk well.

Of course, a lot of this is expected because of the nature of random testing, but I expect we could do better than we currently do.

## Simple failures CsCheck struggles to find

First of all, here's a basic case of a sparse failure pattern on floats.
 
```C#
[Test]
public void IntegerFloatTest()
    => Gen.Float[-10_000, 10_000].Sample(v => Assert.IsFalse(v == MathF.Floor(v)), iter: 100_000);
```

This will typically find a failure in ~10,000 iterations, which isn't too bad, but isn't great. I would guess that failing on integers is a relatively common failure mode, so perhaps it makes sense to bias the generator towards them. I expect there are significantly more complex ways in which the generator could be biased to cover more cases.

`Gen.Float.Special` would catch this particular case, but would easily fail if we added a `v > 5` condition.

Generally, my opinion is that expecting the user to carefully choose their float distribution is bad. Choosing the distribution requires having some type of idea of what distribution is likely to find bugs. Now implementation assumptions are leaking into tests. The default distribution should be carefully chosen to cover as many plausible cases as possible.

Here's another plausible failure pattern.

```C#
[Test]
public void FailSame()
{
    var gen =
        from x in Gen.Float
        from y in Gen.Float
        select (x, y);

    gen.Sample((x, y) => Assert.IsFalse(x == y), iter: 1_000_000);
}
```

This will never be found in 1,000,000 iterations.

The composable nature of the generators makes this quite hard to deal with. They're completely independent, so there's no built in mechanism that could allow them to try the same value case. Of course, we could make some `Point(x, y)` type and a custom generator for it which exercises this case, but that requires the programmer to already have a good guess where a bug may be.

This is also solved by `Gen.Float.Special`, because the bias towards constants means it eventually chooses the same constant multiple times. Again though, this only works if the failure exists at the particular constants `Special` uses.

## Shrinking

I'm also finding it does quite poorly on numerical shrinking compared to other random testing libraries. Intuitively this makes sense; for sparse errors, it's unlikely to find the smallest instance by uniform random sampling. The `IntegerFloatTest` above will almost never find a nice small integer, unless used with a biased generator. Manual shrinkers are better at this.

There's also another issue that's nice and easy to fix. The `Size` of floats is calculated weirdly, and doesn't correspond to any intuitive notion of simplicity, so it doesn't even try to shrink them in any useful way. I've got a simple extension method with a hacky fix for now - just using the magnitude of the value as the size is a big improvement.

## Conclusion

Some of this is just the pitfalls of random testing, but I expect that smart crafting of default distributions could improve these situations significantly. 

I've found CsCheck immensely helpful for certain types of tests! The API is an absolute joy! But for some of these numerical things, it's struggling.

Is it fair to say that numerical testing simply isn't what CsCheck is designed to be good at?

---

I have some more radical ideas which may be able to improve performance on some of the harder problems, including a prototype which drastically improves shrinking performance in my tests. I'll post those in a comment below as this is getting long and messy.

**Comment by AnthonyLloyd** [comment]
I agree that it's probably a good idea to mix some rational values into double and float. I haven't needed to work with floating point equality to see an issue. I'll look to add but it may need to be a major version. You'd need to experiement with your own version of GenFloat to add this in with a range.

I'm not so convinced about mixing in hand shrinkers. I'd have to see an API of how the failure data is then used in the generation. I think this has a path dependency issue that would mean you can't just start from a seed for the shrunk case. My main criticism of tree shrinkers was they can't give you a seed for the shrunk case and they only explore fixed axis. The [version](https://github.com/AnthonyLloyd/CsCheck/blob/master/Comparison.md) example here makes me think float with some rationals would start to work better and probably solve the problem.

**Comment by AnthonyLloyd** [comment]
I've added Gen.Double.WithInt, Gen.Double.WithRational, Gen.Float.WithInt, Gen.Float.WithRational in version 2.13.0 if they are useful. I'm thinking of making these the default in a major version increment at some point. There are many options here though that I'd like to rationalize (pun intended).

**Comment by AnthonyLloyd** [comment]
Working on this for an upcoming 3.0.0-rc1. Double change already in master.
